zigzag-porep.md

```diff
@@ -0,0 +1,529 @@
+
+```go
+type node [32]uint8
+
+// Create and return a binary merkle tree and its root commitment.
+// len(leaves) must be a power of 2.
+func RepHash(leaves []node) ([][]node, node) {
+	rows = [][]node
+	
+	currentRow := leaves
+	for height := 0; len(currentRow) > 1; height += 1 {
+		rows.push(currentRow)
+		var nextRow []node
+
+		for i := 0; i < len(row)/2; i += 2 {
+			left := row[i]
+			right := row[i+1]
+
+			// NOTE: Depending on choice of RepCompress, heightPart may be trimmed to fewer than 8 bits.
+			heightPart := []uint8{height}
+			
+			input1 := append(heightPart, left...)
+			input := append(input1, right...)
+			hashed = RepCompress(input)
+			nextRow = append(nextRow, hashed)
+		}
+
+		currentRow = nextRow
+	}
+    // The tree returned here is just a vector of rows for later use. Its representation is not part of the spec.
+	return rows, currentRow[0]
+}
+
+```
+
+For each of `LAYERS` layers, `l`, perform one __*Layer Replication*__, yielding a replica, tree, and commitment (`CommR_<l>`) per layer:
+
+```go
+let layer_replicas = [LAYERS][nodes]uint8
+let layer_trees = [LAYERS]MerkleTree
+let CommRs = []commitment
+
+let layer = data
+for l in 0..layers {
+	let layer_replica = ReplicateLayer(layer)
+	layer_replicas[l] = layer_replica
+	CommRs[l], layers_trees[l] = RepTree(layer_replica)
+	layer = layer_replica
+}
+```
+
+The replicated data is the output of the final __*Layer Replication*__,`layer_replicas[layers-1]`.
+Set `CommRLast` to be  `CommR_<Layers>`.
+Set `CommRStar` to be `CommRHash(ReplicaID || CommR_0 || CommR_<i> || ... || CommRLast)`.
+
+```go
+Replica := layer_replicas[layers - 1]
+CommRLast :- CommRs[layers-1]
+CommRStar := CommRHash(replicaID, ...CommRs)
+```
+
+## Proof Generation
+
+### Challenge Generation
+
+Calculate `LAYER_CHALLENGES : [LAYERS]uint`: Number of challenges per layer. (This will be passed to the ZigZag circuit proof.)
+
+Derive challenges for each layer (call `derive_challenges()`).
+
+### Layer Challenge Counts
+
+TODO: define `Challenge` (or find existing definition)
+
+```go
+func ChallengesForLayer(challenge Challenge, layer uint) uint {
+
+    switch challenge.Type {
+        case Fixed:
+            return challenge.Count
+        case Tapered:
+            assert(layer < LAYERS)
+            l := (LAYERS - 1) - layer
+            r := 1.0 - TAPER;
+            t := min(l, TAPER_LAYERS)
+
+            totalTaper := pow(r, t)
+
+            calculated := ceil(totalTaper * challenge.count)
+
+            // Although implied by the call to `ceil()` above, be explicit
+            // that a layer cannot contain 0 challenges.
+            max(1, calculated)
+    }
+}
+```
+
+### Challenge Derivation
+
+TODO: define `Domain` (for practical purposes a `uint`) and `LayerChallenges` (or find existing definition).
+
+```go
+func DeriveChallenges(challenges LayerChallenges, layer uint, leaves uint,
+    replicaId Domain, commitment Domain, k uint) []uint {
```

Comment:
we should rename commitment with seed, this would make the spec interactive-porep friendly